### PR TITLE
Cloudbuild worker pools

### PR DESCRIPTION
**Pull request by slevenick**
Add support for cloudbuild worker pools. Fix some base paths that were defaulting to empty strings
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/8985

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_cloudbuild_worker_pool
```

### CODE REVIEW

File: tpgtools/overrides/dataproc/beta/tpgtools_product.yaml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 ## product level overrides
 
 ## SKIP_BASEPATH generation... already generated by magic modules
```
Comment (ScottSuarez) [inline tpgtools/overrides/dataproc/beta/tpgtools_product.yaml:3]: remove outdated comments

File: tpgtools/product.go [hunk 2]
```diff
@@ -46,13 +46,55 @@ func NewProductMetadata(packagePath, productName string) *ProductMetadata {
 	}
 }
 
-func (pm *ProductMetadata) WriteBasePath() bool {
-	po, ok := productOverrides[pm.PackagePath]
-	if !ok {
+func (pm *ProductMetadata) WriteProductBasePath() bool {
```
Comment (ScottSuarez) [inline tpgtools/product.go:49]: should we change this to be `ShouldWriteProductBasePath`?

File: tpgtools/product.go [hunk 3]
```diff
@@ -46,13 +46,55 @@ func NewProductMetadata(packagePath, productName string) *ProductMetadata {
...
-	return !skipBasePath
+	return !bp.Skip
+}
+
+func (pm *ProductMetadata) BasePathNameSnakeUpper() string {
+	bp := pm.ProductBasePathDetails()
+	if bp != nil && bp.BasePathName != ""{
+		return strings.ToUpper(bp.BasePathName)
+	}
+	return pm.ProductNameUpper()
```
Comment (ScottSuarez) [inline tpgtools/product.go:62]: return string.ToUpper(pm.BasePathNameSnake())

File: tpgtools/templates/provider_dcl_client_creation.go.tmpl [hunk 4]
```diff
@@ -46,15 +46,15 @@ func NewDCL{{$pkg.ProductType}}Client(config *Config, userAgent, billingProject
 			dclClientOptions,
 			dclUserAgentOptions,
 			dclLoggerOptions,
-			dcl.WithBasePath(config.{{$pkg.ProductType}}BasePath),
+			dcl.WithBasePath(config.{{$pkg.BasePathName}}BasePath),
```
Comment (ScottSuarez) [inline tpgtools/templates/provider_dcl_client_creation.go.tmpl:49]: perhaps rephrase `BasePathName` to `BasePathId` or `BasePathIdentifier` ? 

hmm im not sure name sounds weird to me

Comment (ScottSuarez) [reply tpgtools/templates/provider_dcl_client_creation.go.tmpl:49]: only nit, feel free to ignore if you disagree

**Comment by ScottSuarez** [comment]
Do you have a lint to the rest api docs ?

File: mmv1/third_party/terraform/tests/resource_cloudbuild_worker_pool_test.go.erb [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func testAccCloudbuildWorkerPool_updated(context map[string]interface{}) string {
+	return Nprintf(`
+resource "google_cloudbuild_worker_pool" "pool" {
+	name = "pool%{random_suffix}"
+	location = "europe-west1"
+	worker_config {
+		disk_size_gb = 101
+		machine_type = "e2-standard-4"
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/tests/resource_cloudbuild_worker_pool_test.go.erb:76]: is this really changeable without deletion ?

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+layout: "google"
+page_title: "Google: google_cloudbuild_worker_pool"
+sidebar_current: "docs-google-cloudbuild-worker-pool"
+description: |-
+  Configuration for custom WorkerPool to run builds
+---
+
+# google\_cloudbuild\_worker\_pool
+
+Definition of custom Cloud Build WorkerPools for running jobs with custom networking.
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:13]:
```suggestion
Definition of custom Cloud Build WorkerPools for running jobs with custom configuration and custom networking.
```

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+-> This resource is not currently public, and requires allow-listing of projects prior to use.
+
+~> **Warning:** This resource is in beta, and should be used with the terraform-provider-google-beta provider.
+See [Provider Versions](https://terraform.io/docs/providers/google/guides/provider_versions.html) for more details on beta resources.
+
+
+## Example Usage
+
+```hcl
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:23]: add network config example

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  (Optional)
+  Network configuration for the `WorkerPool`.
+  
+* `project` -
+  (Optional)
+  The project for the resource
+  
+* `worker_config` -
+  (Optional)
+  Configuration to be used for a creating workers in the `WorkerPool`.
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:60]:
```suggestion
  Configuration for workers in the `WorkerPool`.
```

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  
+
+
+The `network_config` block supports:
+    
+* `peered_network` -
+  (Required)
+  Required. Immutable. The network definition that the workers are peered to. If this section is left empty, the workers will be peered to `WorkerPool.project_id` on the service producer network. Must be in the format `projects/{project}/global/networks/{network}`, where `{project}` is a project number, such as `12345`, and `{network}` is the name of a VPC network in the project. See (https://cloud.google.com/cloud-build/docs/custom-workers/set-up-custom-worker-pool-environment#understanding_the_network_configuration_options)
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:68]:
```suggestion
  Immutable. The network definition that the workers are peered to. If this section is left empty, the workers will be peered to `WorkerPool.project_id` on the service producer network. Must be in the format `projects/{project}/global/networks/{network}`, where `{project}` is a project number, such as `12345`, and `{network}` is the name of a VPC network in the project. See (https://cloud.google.com/cloud-build/docs/custom-workers/set-up-custom-worker-pool-environment#understanding_the_network_configuration_options)
```

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown [hunk 10]
```diff
@@ -0,0 +1,120 @@
+
+---
+subcategory: "Cloud Build"
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:3]: docs look corrupted 

https://github.com/slevenick/magic-modules/blob/worker-pools/mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown (hunk 8 above)
Comment (slevenick) [reply mmv1/third_party/terraform/website/docs/r/cloudbuild_worker_pool.html.markdown:60]: I think the current wording is actually more correct here. Also this comes from the API directly

I don't exactly know how workers are created, but I believe the worker pool contains the config that is used to set up the worker somewhat on-demand.

**Review by ScottSuarez — APPROVED** [review]

**Pull request merged**